File: Spreadsheet/Spreadsheet/Spreadsheet.cs
using SpreadsheetUtilities;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Text.RegularExpressions;
using System.Xml;

namespace SS
{

    /// <summary>
    /// This Class represents the beginning internals of our Spreadsheet program.
    /// It utilizes a Dependency graph to keep track of cell dependencies, as well
    /// as using a Formula class to evaluate formulas.
    /// @Author Tarik Vu u0759288 
    /// 9/25/20
    ///  
    /// EDIT: As of 9/29/20 we are on the PS5 Branch! - Tarik Vu
    /// </summary>
    public class Spreadsheet : AbstractSpreadsheet
    {
        /// <summary>
        /// Our backing data structure (dictionary) that will contain all of our cells
        /// using their cell name as their key.
        /// </summary>
        private IDictionary<string, Cell> Cells;

        /// <summary>
        /// Our Dependency graph that keeps track of each Cell's Dependency, also used in finding 
        /// circular exceptions.
        /// </summary>
        private DependencyGraph depGraph;

        /// <summary>
        /// Indicates whether or not this SS has changed from either being saved or a new 
        /// if a new cell was added.
        /// </summary>
        public override bool Changed { get; protected set; }


        /// <summary>
        /// Constructor that takes in no arguements to construct an empty spreadsheet.
        /// Using base() we can have lamda's to determine the default returns for our delegates.
        /// In a zero argument construction we are to have isValid always be true, normalize to always
        /// return the string as it was, and our version to be set to the string "default".
        /// 
        /// Validity delegate:  Used to check if a variable name is valid.
        /// Normalize delegate: Used to Standardize (normalize) variable names.
        /// Version:            The spreadsheet's version.
        /// </summary>
        public Spreadsheet()
            : base(v => true, n => n, "default")
        {
            Cells = new Dictionary<string, Cell>();
            depGraph = new DependencyGraph();
            Changed = false;
        }

        /// <summary>
        /// Constructor that takes in three arguements to construct an empty spreadsheet.
        /// 
        /// Validity delegate:  Used to check if a variable name is valid.
        /// Normalize delegate: Used to Standardize (normalize) variable names.
        /// Version:            The spreadsheet's version.
        /// </summary>
        public Spreadsheet(Func<string, bool> isValid, Func<string, string> normalize, string version)
            : base(isValid, normalize, version)
        {
            Cells = new Dictionary<string, Cell>();
            depGraph = new DependencyGraph();
            Changed = false;
        }

        /// <summary>
        /// Constructor that takes in four arguements to construct a spreadsheet.
        /// 
        /// Path:               String representing path to a file after being saved.
        /// Validity delegate:  Used to check if a variable name is valid.
        /// Normalize delegate: Used to Standardize (normalize) variable names.
        /// Version:           The spreadsheet's version.
        /// 
        /// This constructor utilizes XMLDemo's ReadXML method provided by professor kopta.
        /// </summary>
        public Spreadsheet(string path, Func<string, bool> isValid, Func<string, string> normalize, string version)
            : base(isValid, normalize, version)
        {
            Cells = new Dictionary<string, Cell>();
            depGraph = new DependencyGraph();

            // Useing path to get our desired SS
            string filename = path;
            string cellName = "";
            try
            {
                // Create an XmlReader 
                using (XmlReader reader = XmlReader.Create(filename))
                {
                    while (reader.Read())
                    {
                        if (reader.IsStartElement())
                        {
                            switch (reader.Name)
                            {
                                case "spreadsheet":

                                    if (!(this.Version.Equals(reader["version"])))
                                        throw new SpreadsheetReadWriteException("Incorrect version");
                                    break; // no more direct info to read for Spreadsheet

                                case "cell":
                                    break; // no more direct info to read on Cell

                                case "name":
                                    reader.Read();
                                    cellName = reader.Value; // Grab cell name
                                    break;

                                case "contents":
                                    reader.Read(); // reader value is now cell contents.
                                    SetContentsOfCell(cellName, reader.Value);
                                    break;
                            }
                        }
                        else // If it's not a start element, it's probably an end element
                        {
                        }
                    }
                }
            }
            catch
            {
                throw new SpreadsheetReadWriteException("Error loading in spreadsheet " + path);
            }

            Changed = false;
        }



        public override object GetCellContents(string name)
        {
            // Normalize our cell name
            string cellName = Normalize(name);

            // Check for name 
            CheckCellName(cellName);

            // Cell doesn't Exist
            if (!Cells.ContainsKey(cellName))
                return "";

            Cell temp = Cells[cellName];
            return temp.GetContents();
        }

        public override object GetCellValue(string name)
        {
            // Normalize our cell name
            string cellName = Normalize(name);

            // If name is null or invalid.
            CheckCellName(cellName);
            if (Cells.ContainsKey(cellName))
            {
                Cell temp = Cells[cellName];


                return temp.GetValue();
            }

            return "";

        }


        public override IEnumerable<string> GetNamesOfAllNonemptyCells()
        {
            // Check if we have any cells to return
            if (Cells.Values.Count > 0)
                foreach (Cell c in Cells.Values)
                    if (!c.isEmpty)
                        yield return c.name;

            // Return empty
            yield break;
        }

        public override string GetSavedVersion(string filename)
        {
            try
            {
                // Create an XmlReader 
                using (XmlReader reader = XmlReader.Create(filename))
                {
                    while (reader.Read())
                    {
                        if (reader.IsStartElement())
                        {
                            switch (reader.Name)
                            {
                                case "spreadsheet":
                                    return reader["version"];   // Returning Version                                                            
                            }
                        }
                        else // If it's not a start element, it's probably an end element
                        {
                        }
                    }
                }
                // the case for spreadsheet never happened, there was a bad XML file:
                throw new SpreadsheetReadWriteException("unable to find version for. " + filename);
            }
            catch
            {
                throw new SpreadsheetReadWriteException("Error loading in spreadsheet " + filename + ". Unable" +
                    "to retrive version");
            }
        }

        /// <summary>
        /// This Save method is laid out very similarly to the XML demo provided with an added try - catch
        /// per method specifications.
        /// </summary>
        /// <param name="filename">Name of the file.</param>
        public override void Save(string filename)
        {
            // Setting up XML reader
            XmlWriterSettings settings = new XmlWriterSettings();
            settings.Indent = true;
            settings.IndentChars = "  ";

            try
            {
                // Create an XmlWriter 
                using (XmlWriter writer = XmlWriter.Create(filename, settings))
                {

                    writer.WriteStartDocument();
                    writer.WriteStartElement("spreadsheet");

                    // Version (attribute) of this spreadhsheet
                    writer.WriteAttributeString("version", this.Version);

                    // write the Cells themselves
                    foreach (Cell c in Cells.Values)
                        c.WriteXml(writer);

                    writer.WriteEndElement(); // Ends the spreadsheet block
                    writer.WriteEndDocument();

                }
            }
            catch
            {
                throw new SpreadsheetReadWriteException("Error saving SS");
            }

            Changed = false;
        }

        protected override IList<string> SetCellContents(string name, double number)
        {
            // Adding as a new cell
            if (!(Cells.ContainsKey(name)))
                Cells.Add(name, new Cell(name, number));

            // overwriting cell 
            else
                Cells[name] = new Cell(name, number);

            // Recalculate dependents
            RecalculateDependents(name);

            List<string> toReturn = GetCellsToRecalculate(name).ToList();
            Changed = true;

            return toReturn;

        }

        protected override IList<string> SetCellContents(string name, string text)
        {

            // Add new empty cell if needed
            if (!(Cells.ContainsKey(name)))
                Cells.Add(name, new Cell(name, text));

            // overwriting a cell with an empty string removes it (piazza @670)
            else if (text.Equals(""))
                Cells.Remove(name);

            // overwriting cell 
            else
                Cells[name] = new Cell(name, text);

            List<string> toReturn = GetCellsToRecalculate(name).ToList();

            // Cells that depended on Cells[name] will now be formula error after recalc
            RecalculateDependents(name);

            Changed = true;

            return toReturn;
        }

        protected override IList<string> SetCellContents(string name, Formula formula)
        {

            foreach (string var in formula.GetVariables())
            {

                if (!IsValid(var))
                    throw new FormulaFormatException("invalid variable " + var);

            }

            // Try Catch in case of Circular Exception
            try
            {
                // Adding dependencies for the variables within formula to name
                foreach (string var in formula.GetVariables())
                    depGraph.AddDependency(var, name);

                // Our list to return as well as check for CircularException
                List<string> toReturn = GetCellsToRecalculate(name).ToList();

                // Adding as a new Cell and grabbing it's value.
                object value;
                try
                {
                    value = formula.Evaluate(lookup);
                }
                catch (ArgumentException)
                {
                    value = new FormulaError();
                }

                if (!(Cells.ContainsKey(name)))
                    Cells.Add(name, new Cell(name, formula, value));

                // Overwriting the Cell to a formula
                else
                    Cells[name] = new Cell(name, formula, value);

                // Recalculate dependents
                RecalculateDependents(name);

                Changed = true;
                return toReturn;
            }

            // Resetting dependencies SS is unchanged.
            catch (CircularException)
            {
                foreach (string var in formula.GetVariables())
                    depGraph.RemoveDependency(var, name);

                // Then we throw the Exception.
                throw new CircularException();
            }

        }

        public override IList<string> SetContentsOfCell(string name, string content)
        {
            // Normalize our cell name
            string cellName = Normalize(name);

            // Name and content check here, removed from other add methods.
            CheckCellName(cellName);
            if (content is null)
                throw new ArgumentNullException();

            // If content parses as double set contents to double.
            double doubleContent;
            bool isDouble = double.TryParse(content, out doubleContent);
            if (isDouble)
                return SetCellContents(cellName, doubleContent);

            // Formula case needs a string after '=', otherwise just the string "="
            if (content.StartsWith("="))
            {
                string remainder = content.Substring(1, content.Length - 1);

                // Normalize the variables in our formula
                remainder = Normalize(remainder);

                // FormulaFormatException would be thrown here:
                Formula f = new Formula(remainder);
                return SetCellContents(cellName, f);
            }

            // Setting as string:
            return SetCellContents(cellName, content);
        }

        protected override IEnumerable<string> GetDirectDependents(string name)
        {
            // Check for any Dependents to begin with
            if (depGraph.HasDependents(name))
            {
                return depGraph.GetDependents(name);
            }
            // Otherwise we return a new empty list
            return new List<string>();
        }


        // Private helper methods for Spreadsheet

        /// <summary>
        /// Private helper method that uses our dictionary "Cells" in order to 
        /// lookup a cell's value for evaluation.
        /// This method will throw an arguement exception if:
        /// - A Cell is empty (value is empty string)
        /// - A cell's value is a string / formula error
        /// This in turn will cause an evaluation in the Formula Class of this cell
        /// to result in a return of a FormulaError.
        /// </summary>
        /// <param name="name">Cell Name</param>
        /// <returns>Cell's double value.</returns>
        private double lookup(string name)
        {
            // Looking up a cell that hasnt been added yet, add an empty one.
            if (!Cells.ContainsKey(name))
                Cells.Add(name, new Cell(name, ""));

            Cell cell = Cells[name];

            object value = cell.GetValue();

            // We have a string or FormulaError as the value of our cell.
            if (value is string || value is FormulaError)
                throw new ArgumentException();

            // Else we know our cell value is a double and we can return it as such.
            return (double)value;
        }

        /// <summary>
        /// Private helper method to recalculate the dependents of 
        /// this Cell[name].  This method is called if "name" has dependents that
        /// depend on it's value. And is called in the case where the contents of this
        /// cell were either a double or formula.
        /// </summary>
        /// <param name="name">Name of cell that has dependents</param>
        private void RecalculateDependents(string name)
        {
            // Updating the values
            foreach (string s in GetCellsToRecalculate(name))
            {
                // Check if that cell was removed already, if so skip and just recalc dependents.
                if (Cells.ContainsKey(name))
                {
                    object value;
                    Cell cur = Cells[s];
                    Formula formula = cur.formula;
                    try
                    {
                        // Only recalculating if needed
                        if (!(formula is null))
                        {
                            // Try to evaluate the dependent
                            value = formula.Evaluate(lookup);

                            // Adding the cell back in with its new value.
                            Cells[s] = new Cell(s, formula, value);
                        }
                    }

                    // Dependent (still) couldnt evaluate, give it an Error.
                    catch (ArgumentException)
                    {
                        Cells[s] = new Cell(s, formula, new FormulaError());
                    }
                }
            }
        }

        /// <summary>
        /// UPDATED FOR PS5 DEFINITION
        /// Private helper method called whenever a cell is added.
        /// that check's the Cells Name for validity.
        /// 
        /// Variables for a Spreadsheet are valid if and only if they are one or more letters followed
        /// by one or more digits (numbers) this must now be enforced by the spreadsheet.
        /// 
        /// Throws InvalidNameException when a cell name condition is not met.
        /// </summary>
        /// <param name="name">Cell name to check</param>
        private void CheckCellName(string name)
        {
            // Checking Cell name validity with our delegate:
            if (IsValid(name))
            {

                // Null Check 
                if (name is null || name.Equals(""))
                    throw new InvalidNameException();

                // Start of variable must be letter 
                if (!(Char.IsLetter(name[0])))
                    throw new InvalidNameException();

                // First Letter must be followed by letters or nums
                if (name.Length < 2)
                    throw new InvalidNameException();

                // Regex to determine remaining string Letters followed by numbers
                string remaining = name.Substring(1, name.Length - 1);

                string pattern = "^[a-zA-Z]*[0-9]+$";
                if (!Regex.IsMatch(remaining, pattern))
                    throw new InvalidNameException();
            }
            else
            {
                throw new InvalidNameException();
            }
        }

        /// <summary>
        /// Our Cell Class that will be used in Spreadsheet.  Every time a new cell is created
        /// it's contents are defaulted to an empty string.  
        ///
        /// A Cell Also contains a contents and a value.
        /// Contents can be: String, double, or Formula.
        /// If contents are empty string "" the cell is considered empty.       
        /// Value can be: String, double, or FormulaError.
        /// 
        /// If Contents = String, Value is that string.
        /// If Contents = double, Value is that double.
        /// If Contents = Formula, Value is double or Formula Error as reported by
        /// the evaluate method of Formula Class.
        /// 
        /// our private doubles and strings are used to represent our 
        /// different contents.
        /// 
        /// Values are to be implemented during ps5
        /// value: This private object is to be used to either hold a double
        /// or a Formula error.  It is created when a cell is created using the formula 
        /// constructor.  
        /// </summary>
        private class Cell
        {
            /// <summary> Double Contents of cell (if any)  </summary>
            private double conDouble;

            /// <summary> String Contents of cell (if any) </summary>
            private string conString;

            /// <summary> Formula Contents of cell (if any) </summary>
            public Formula formula { get; private set; }

            // ADDED FOR PS5
            /// <summary> Formula Value to represent double or Formula error </summary>
            private object value;


            /// <summary> The name of this cell, Made immutable with private set. </summary>
            public string name { get; private set; }

            /// <summary> Dictates whether or not this cell is empty. </summary>
            public bool isEmpty { get; private set; }

            /// <summary> Creating a new cell with a name, and it's input double.</summary>
            public Cell(string name, double number)
            {
                this.name = name;
                this.conDouble = number;
                isEmpty = false;
            }

            /// <summary> Creating a new cell with a name, and it's input String.</summary>
            public Cell(string name, string text)
            {
                this.name = name;
                this.conString = text;

                // If the text was an empty string, the cell is empty.
                if (text.Equals(""))
                    isEmpty = true;
                else
                    isEmpty = false;
            }

            /// <summary>
            /// Creates a Cell with the given name, formula, and value.
            /// Value is determined when the cell is first added and can either be
            /// a double, or a FormulaError.
            /// </summary>
            /// <param name="name">Cell's Name</param>
            /// <param name="formula">Cell's Formula</param>
            /// <param name="value">Either a double or formula error.</param>
            public Cell(string name, Formula formula, object value)
            {
                this.name = name;
                this.formula = formula;
                this.value = value;
                isEmpty = false;
            }

            /// <summary>
            /// Returns the contents of this cell by examining which variable was instantiated
            /// using the appropriate constructor.  
            /// </summary>
            /// <returns>Returns the contents of this cell (string, double, or Formula)</returns>
            public object GetContents()
            {
                // Contents were string
                if (!(conString is null))
                    return conString;

                // Contents were Formula
                if (!(formula is null))
                    return formula;

                // Contents were double
                return conDouble;
            }

            /// <summary>
            /// Returns the Value of this cell by examining which variable was instantiated
            /// using the appropriate constructor.  
            /// In the case that the Contents of this cell were a formula, we are to evaluate said
            /// formula to return either a double or a formula error.
            /// </summary>
            /// <returns>Returns the contents of this cell (string, double, or Formula)</returns>
            public object GetValue()
            {
                // Contents were string
                if (!(conString is null))
                    return conString;

                // Contents were Formula return value (double or FormulaError)
                if (!(formula is null))
                    return value;

                // Contents were double
                return conDouble;
            }

            /// <summary>
            /// This method was inspired by the XML Demo provided by Prof. Kopta.  
            /// It is used with an XMLWriter when saving a spreadsheet.  And writes down 
            /// the data of this specific cell including the cell's name and contents.  The
            /// value is not recorded to save on space and is simply reconstructed when loading
            /// in a spreadsheet.
            /// When writing a Formula contents, we prepend "=" as per the save method's specifications.
            /// </summary>
            /// <param name="writer">XML writer used to record this given cell.</param>
            internal void WriteXml(XmlWriter writer)
            {
                writer.WriteStartElement("cell");
                // We use a shortcut to write an element with a single string
                writer.WriteElementString("name", name);

                // Determining the appropriate contents
                string contents;

                // Contents were string
                if (!(conString is null))
                    contents = conString;

                // Contents were Formula 
                else if (!(formula is null))
                    contents = formula.ToString().Insert(0, "=");

                // Contents were double
                else
                    contents = conDouble.ToString();

                writer.WriteElementString("contents", contents);
                writer.WriteEndElement(); // Ends the Cell
            }
        }
    }
}

File: Server/Server/Spreadsheet.cpp
#include "Spreadsheet.h"


/// <summary>
/// This Class represents the beginning internals of our Spreadsheet program.
/// It utilizes a Dependency graph to keep track of cell dependencies, as well
/// as using a Formula class to evaluate formulas.
/// @Author Tarik Vu u0759288 
/// 9/25/20
///  
/// EDIT: As of 9/29/20 we are on the PS5 Branch! - Tarik Vu
/// 
/// Edit: 4/21/2021 Modifed for C++ for 3505 final.
/// @Author "It Should work now"
/// </summary>

/// <summary>
/// default constructor
/// </summary>
Spreadsheet::Spreadsheet() : spreadsheet_name("spwedsweet UwU"), changed(false)
{
}


Spreadsheet::Spreadsheet(std::string& name) : spreadsheet_name(name), changed(false)
{
}

//std::map<Cell, std::string> Spreadsheet::get_spreadsheet_contents() {
//	std::map<Cell, std::string> content; 
//
//	return content; 
//}

std::string Spreadsheet::get_spreadsheet_name() { return spreadsheet_name;  }

// MODIFIED PROTECTION FOR PS5
/// <summary>
/// The contents of the named cell becomes text.  The method returns a
/// list consisting of name plus the names of all other cells whose value depends, 
/// directly or indirectly, on the named cell. The order of the list should be any
/// order such that if cells are re-evaluated in that order, their dependencies 
/// are satisfied by the time they are evaluated.
/// 
/// For example, if name is A1, B1 contains A1*2, and C1 contains B1+A1, the
/// list {A1, B1, C1} is returned.
/// </summary>
std::list<std::string> Spreadsheet::set_cell_content(std::string& cellName, double number)
{
	std::string num = std::to_string(number);
	set_generic_content(cellName, num);

	//updates the dependencies
	std::vector<std::string> s;
	cell_graph.replace_dependents(cellName, s);

	std::list<std::string> list = get_cells_to_recalculate(cellName);

	for (std::string s : list)
	{
		std::string new_cont(update_value(s));
		cells_map[s].set_cell_content(new_cont);
	}

	return list;
}

/// <summary>
/// The contents of the named cell becomes text.  The method returns a
/// list consisting of name plus the names of all other cells whose value depends, 
/// directly or indirectly, on the named cell. The order of the list should be any
/// order such that if cells are re-evaluated in that order, their dependencies 
/// are satisfied by the time they are evaluated.
/// 
/// For example, if name is A1, B1 contains A1*2, and C1 contains B1+A1, the
/// list {A1, B1, C1} is returned.
/// </summary>
std::list<std::string> Spreadsheet::set_cell_content(std::string& cellName, std::string text)
{
	//updates the dependencies
	std::vector<std::string> vec;
	cell_graph.replace_dependents(cellName, vec);

	std::list<std::string> list = get_cells_to_recalculate(cellName);

	if (text.empty() && cells_map.contains(cellName))
	{
		cells_map.erase(cellName);
	}
	else if (!text.empty())
	{
		set_generic_content(cellName, text);

		for (auto i : list)
		{
			if (i == cellName)
				continue;

			std::string s("");
			s = i; //THERE IS AN ERROR HERE
			std::cout << i << std::endl;
			std::cout << s << std::endl;
			std::string new_cont(update_value(s));
			cells_map[s].set_cell_content(new_cont); // this is updating the current cell twice
		}
	}

	return list;
}

/// <summary>
/// If content is null, throws an ArgumentNullException.
/// 
/// Otherwise, if name is null or invalid, throws an InvalidNameException.
/// 
/// Otherwise, if content parses as a double, the contents of the named
/// cell becomes that double.
/// 
/// Otherwise, if content begins with the character '=', an attempt is made
/// to parse the remainder of content into a Formula f using the Formula
/// constructor.  There are then three possibilities:
/// 
///   (1) If the remainder of content cannot be parsed into a Formula, a 
///       SpreadsheetUtilities.FormulaFormatException is thrown.
///       
///   (2) Otherwise, if changing the contents of the named cell to be f
///       would cause a circular dependency, a CircularException is thrown,
///       and no change is made to the spreadsheet.
///       
///   (3) Otherwise, the contents of the named cell becomes f.
/// 
/// Otherwise, the contents of the named cell becomes content.
/// 
/// If an exception is not thrown, the method returns a list consisting of
/// name plus the names of all other cells whose value depends, directly
/// or indirectly, on the named cell. The order of the list should be any
/// order such that if cells are re-evaluated in that order, their dependencies 
/// are satisfied by the time they are evaluated.
/// 
/// For example, if name is A1, B1 contains A1*2, and C1 contains B1+A1, the
/// list {A1, B1, C1} is returned.
/// </summary>
std::list<std::string> Spreadsheet::set_contents_of_cell(std::string& name, std::string& content)
{
	if (content.empty())
		throw std::invalid_argument("NO EMPTY STRINGS");

	std::string normalized = normalize(name);
	name_check(normalized);

	changed = true;
	stack_changes.push(normalized);

	//three cases possible content is either a:
	//number
	try
	{
		double d = std::stod(normalized);
		return set_cell_content(normalized, d);
	}
	catch (std::invalid_argument e)
	{
		//formula
		if (content.size() > 0 && content.at(0) == '=')
		{
			std::string s = content.substr(1);
			Formula formula(s);

			return set_cell_content(normalized, formula);
		}
		//text
		else
		{
			return set_cell_content(normalized, content);
		}
	}
}

/// <summary>
/// undoes the last change to the spreadsheet 
/// </summary>
/// <returns>the new content of the cell</returns>
std::string Spreadsheet::undo(std::string& name)
{
	name = stack_changes.peek();
	name = normalize(name);
	return revert(stack_changes.pop(), true);
}

/// <summary>
/// Reverts a specific cell back to its previous value
/// </summary>
/// <param name="name">name of the cell to be reverted</param>
/// <returns>the new content of the cell</returns>
std::string Spreadsheet::revert(const std::string& name, bool is_undo)
{
	try
	{
		if (is_undo)
		{
			return cells_map.at(name).revert();
		}
		else
		{	
			//should add to the changes stack
			return cells_map.at(name).revert();
		}
	}
	catch (std::out_of_range e)
	{
		std::cout << "revert on nonexistent cell" << std::endl;
		return "";
	}
}

/// <summary>
/// If changing the contents of the named cell to be the formula would cause a 
/// circular dependency, throws a CircularException, and no change is made to the spreadsheet.
/// 
/// Otherwise, the contents of the named cell becomes formula. The method returns a
/// list consisting of name plus the names of all other cells whose value depends,
/// directly or indirectly, on the named cell. The order of the list should be any
/// order such that if cells are re-evaluated in that order, their dependencies 
/// are satisfied by the time they are evaluated.
/// 
/// For example, if name is A1, B1 contains A1*2, and C1 contains B1+A1, the
/// list {A1, B1, C1} is returned.
/// </summary>
std::list<std::string> Spreadsheet::set_cell_content(std::string& cellName, Formula formula)
{
	//adds the variables to the dependency graph
	std::vector<std::string> variables = formula.get_variables();
	cell_graph.replace_dependents(cellName, variables);

	//checks if the graph is valid
	try
	{
		std::list<std::string> list = get_cells_to_recalculate(cellName);

		//an "=" at the begining of a string is how we determine if something is a formula
		std::string f = "=" + formula.to_string();
		set_generic_content(cellName, f);//Fs in the chat bois

		for (std::string s : list)
		{
			std::string new_cont(update_value(s));
			cells_map[s].set_cell_content(new_cont);
		}

		return list;
	}
	catch (CircularException& e)
	{
		for (std::string s : variables)
			cell_graph.remove_dependency(cellName, s);

		throw CircularException("HOLY SHIT A WILD CIRCULAR DEPENDENCY APPEARED");
	}

	return std::list<std::string>();
}


void Spreadsheet::set_spreadsheet_name(const std::string& new_name) {
	spreadsheet_name = new_name;
}

/// <summary>
/// Otherwise, returns the contents (as opposed to the value) of the named cell.  The return
/// value should be either a string, a double, or a Formula.
/// </summary>
std::string Spreadsheet::get_cell_contents(const std::string& name)
{
	if (cells_map.contains(name))
	{
		std::string s(cells_map[name].get_cell_content());
		return s;
	}
	else
	{
		return "";
	}
}

/// <summary>
/// Enumerates the names of all the non-empty cells in the spreadsheet.
/// </summary>
std::vector<std::string> Spreadsheet::get_nonempty_cells()
{
	std::vector<std::string> ret;

	for (auto s : cells_map)
	{
		ret.push_back(s.first);
	}

	return ret;
}


/// <summary>
/// helper method for the set methods
/// adds an item if there is not one or sets it if there is one already
/// </summary>
/// <param name="name">name of the cell</param>
/// <param name="content">content to be set</param>
void Spreadsheet::set_generic_content(std::string& name, std::string& content)
{
	std::string s(name);


	if (cells_map.contains(name))
	{
		cells_map[name].set_cell_content(content);
	}
	else
	{
		Cell c(name, content);
		cells_map.insert({ name, c });//after this line the cell loses its name
	}
}

/// <summary>
/// A convenience method for invoking the other version of GetCellsToRecalculate
/// with a singleton set of names.  See the other version for details.
/// </summary>
std::list<std::string> Spreadsheet::get_cells_to_recalculate(const std::string& name)
{
	std::vector<std::string> vec;
	vec.push_back(name);

	return get_cells_to_recalculate(vec);
}

/// <summary>
/// Requires that names be non-null.  Also requires that if names contains s,
/// then s must be a valid non-null cell name.
/// 
/// If any of the named cells are involved in a circular dependency,
/// throws a CircularException.
/// 
/// Otherwise, returns an enumeration of the names of all cells whose values must
/// be recalculated, assuming that the contents of each cell named in names has changed.
/// The names are enumerated in the order in which the calculations should be done.  
/// 
/// For example, suppose that 
/// A1 contains 5
/// B1 contains 7
/// C1 contains the formula A1 + B1
/// D1 contains the formula A1 * C1
/// E1 contains 15
/// 
/// If A1 and B1 have changed, then A1, B1, and C1, and D1 must be recalculated,
/// and they must be recalculated in either the order A1,B1,C1,D1 or B1,A1,C1,D1.
/// The method will produce one of those enumerations.
/// 
/// Please note that this method depends on the abstract GetDirectDependents.
/// It won't work until GetDirectDependents is implemented correctly.
/// </summary>
std::list<std::string> Spreadsheet::get_cells_to_recalculate(std::vector<std::string>& names)
{
	std::list<std::string> changed;
	std::unordered_set<std::string> visited;

	for (std::string name : names)
	{
		if (!visited.contains(name))
		{
			visit(name, name, visited, changed);
		}
	}

	return changed;
}

/// <summary>
/// Returns an enumeration, without duplicates, of the names of all cells whose
/// values depend directly on the value of the named cell.  In other words, returns
/// an enumeration, without duplicates, of the names of all cells that contain
/// formulas containing name.
/// 
/// For example, suppose that
/// A1 contains 3
/// B1 contains the formula A1 * A1
/// C1 contains the formula B1 + A1
/// D1 contains the formula B1 - C1
/// The direct dependents of A1 are B1 and C1
/// </summary>
std::vector<std::string> Spreadsheet::get_direct_dependents(std::string& name)
{
	return cell_graph.get_dependees(name);
}

/// <summary>
/// A helper for the GetCellsToRecalculate method.
/// </summary>
void Spreadsheet::visit(const std::string& start, std::string& name, std::unordered_set<std::string>& visited, std::list<std::string>& changed)
{
	visited.insert(name);
	for (std::string n : get_direct_dependents(name))
	{
		if (n == start)
		{
			throw CircularException("HOLY SHIT IT'S A CIRCULAR DEPENDENCY");
		}
		else if (!visited.contains(n))
		{
			visit(start, n, visited, changed);
		}
	}

	changed.push_front(name);
}

/// <summary>
/// calculates the value of the cell
/// should acount for the return type (double, string, Formula)
/// </summary>
/// <param name="name"></param>
/// <returns>the value of the calculated cell</returns>
std::string Spreadsheet::update_value(std::string& name)
{
	std::string s(get_cell_contents(name));
	Cell cell(name, s);

	if (cell.is_formula())
	{
		std::string s(cell.get_cell_content());
		Formula f(s);

		std::string evaluate(f.evaluate().get_content());
		//Cell c(name, evaluate);
		return f.evaluate().get_content();
	}
	else
	{
		return s;
	}
}

/// <summary>
/// Normalizes the string 
/// this means converting it to lowercase
/// </summary>
/// <param name="s"></param>
/// <returns>a normalized string</returns>
std::string Spreadsheet::normalize(const std::string& s)
{
	std::locale loc;
	std::string str(s);

	for (std::string::size_type i = 0; i < s.length(); ++i)
		str[i] = std::tolower(s[i], loc);

	return str;
}

/// <summary>
/// checks a string to see if it is a variable
/// </summary>
/// <param name="s"></param>
/// <returns>true if the string is a variable</returns>
bool Spreadsheet::name_check(const std::string& s)
{
	std::string varPattern = "[a-zA-Z]?[0-9]*";

	return std::regex_match(s, std::regex(varPattern));
}